### ISSUE TITLE
Freenode bot stalls and gets disconnected

### ISSUE THREAD

**Issue by jaraco** [yougov/pmxbot#65]
The instance of pmxbot hosted by dcpython.org and connected to freenode finds itself unable to sustain a connection for more than a few minutes.

Watching the debug logs, I see:

```
May 24 14:11:46 kafka pmxbot[19940]: TO SERVER: PING keep-alive
May 24 14:11:46 kafka pmxbot[19940]: FROM SERVER: :adams.freenode.net PONG adams.freenode.net :keep-alive
May 24 14:11:46 kafka pmxbot[19940]: _dispatcher: all_raw_messages
May 24 14:11:46 kafka pmxbot[19940]: command: pong, source: adams.freenode.net, target: adams.freenode.net, arguments: ['keep-alive'], tags: None
May 24 14:11:46 kafka pmxbot[19940]: _dispatcher: pong
May 24 14:12:10 kafka pmxbot[19940]: FROM SERVER: :Mmike!~mario@pagar.splivalo.hr JOIN #mongodb
May 24 14:12:10 kafka pmxbot[19940]: _dispatcher: all_raw_messages
May 24 14:12:10 kafka pmxbot[19940]: command: join, source: Mmike!~mario@pagar.splivalo.hr, target: #mongodb, arguments: [], tags: None
May 24 14:12:10 kafka pmxbot[19940]: _dispatcher: join
```

Then at 14:16, the bot is 'quit' from the channels with a "Ping timeout: 246 seconds".

I've seen MongoDB timeout exceptions in the logs (previously), with many (if not all) associated with these on_join or on_leave events. There are 487k records in the `rolls` table for this bot instance, so that may be implicated.

Indeed, waiting long enough, the operation eventually fails:

```
May 24 14:27:41 kafka pmxbot[19940]: Error in <pmxbot.core.JoinHandler object at 0x7f54241a10f0>
May 24 14:27:41 kafka pmxbot[19940]: Traceback (most recent call last):
May 24 14:27:41 kafka pmxbot[19940]:   File "/opt/pmxbot/lib/python3.6/site-packages/pymongo/pool.py", line 450, in receive_message
May 24 14:27:41 kafka pmxbot[19940]:     self.sock, operation, request_id, self.max_message_size)
May 24 14:27:41 kafka pmxbot[19940]:   File "/opt/pmxbot/lib/python3.6/site-packages/pymongo/network.py", line 137, in receive_message
May 24 14:27:41 kafka pmxbot[19940]:     header = _receive_data_on_socket(sock, 16)
May 24 14:27:41 kafka pmxbot[19940]:   File "/opt/pmxbot/lib/python3.6/site-packages/pymongo/network.py", line 164, in _receive_data_on_socket
May 24 14:27:41 kafka pmxbot[19940]:     chunk = sock.recv(length)
May 24 14:27:41 kafka pmxbot[19940]: TimeoutError: [Errno 110] Connection timed out
May 24 14:27:41 kafka pmxbot[19940]: During handling of the above exception, another exception occurred:
May 24 14:27:41 kafka pmxbot[19940]: Traceback (most recent call last):
May 24 14:27:41 kafka pmxbot[19940]:   File "/opt/pmxbot/lib/python3.6/site-packages/pmxbot/irc.py", line 150, in on_join
May 24 14:27:41 kafka pmxbot[19940]:     handler.attach(locals())()
May 24 14:27:41 kafka pmxbot[19940]:   File "/opt/pmxbot/lib/python3.6/site-packages/pmxbot/notify.py", line 77, in notifier
May 24 14:27:41 kafka pmxbot[19940]:     for msg in Notify.store.lookup(nick):
May 24 14:27:41 kafka pmxbot[19940]:   File "/opt/pmxbot/lib/python3.6/site-packages/pmxbot/notify.py", line 51, in lookup
May 24 14:27:41 kafka pmxbot[19940]:     for msg in res:
May 24 14:27:41 kafka pmxbot[19940]:   File "/opt/pmxbot/lib/python3.6/site-packages/pymongo/cursor.py", line 1114, in next
May 24 14:27:41 kafka pmxbot[19940]:     if len(self.__data) or self._refresh():
May 24 14:27:41 kafka pmxbot[19940]:   File "/opt/pmxbot/lib/python3.6/site-packages/pymongo/cursor.py", line 1036, in _refresh
May 24 14:27:41 kafka pmxbot[19940]:     self.__collation))
May 24 14:27:41 kafka pmxbot[19940]:   File "/opt/pmxbot/lib/python3.6/site-packages/pymongo/cursor.py", line 873, in __send_message
May 24 14:27:41 kafka pmxbot[19940]:     **kwargs)
May 24 14:27:41 kafka pmxbot[19940]:   File "/opt/pmxbot/lib/python3.6/site-packages/pymongo/mongo_client.py", line 905, in _send_message_with_response
May 24 14:27:41 kafka pmxbot[19940]:     exhaust)
May 24 14:27:41 kafka pmxbot[19940]:   File "/opt/pmxbot/lib/python3.6/site-packages/pymongo/mongo_client.py", line 916, in _reset_on_error
May 24 14:27:41 kafka pmxbot[19940]:     return func(*args, **kwargs)
May 24 14:27:41 kafka pmxbot[19940]:   File "/opt/pmxbot/lib/python3.6/site-packages/pymongo/server.py", line 136, in send_message_with_response
May 24 14:27:41 kafka pmxbot[19940]:     response_data = sock_info.receive_message(1, request_id)
May 24 14:27:41 kafka pmxbot[19940]:   File "/opt/pmxbot/lib/python3.6/site-packages/pymongo/pool.py", line 452, in receive_message
May 24 14:27:41 kafka pmxbot[19940]:     self._raise_connection_failure(error)
May 24 14:27:41 kafka pmxbot[19940]:   File "/opt/pmxbot/lib/python3.6/site-packages/pymongo/pool.py", line 550, in _raise_connection_failure
May 24 14:27:41 kafka pmxbot[19940]:     _raise_connection_failure(self.address, error)
May 24 14:27:41 kafka pmxbot[19940]:   File "/opt/pmxbot/lib/python3.6/site-packages/pymongo/pool.py", line 211, in _raise_connection_failure
May 24 14:27:41 kafka pmxbot[19940]:     raise AutoReconnect(msg)
May 24 14:27:41 kafka pmxbot[19940]: pymongo.errors.AutoReconnect: mongodb.jaraco.com:27017: [Errno 110] Connection timed out
```

**Comment by jaraco** [comment]
Aha. It's in the Notify store that the stall is occurring.

**Comment by jaraco** [comment]
That database collection is empty, so any queries to it should be mighty fast, not failing to connect for 15 minutes. I'm going to try some MongoDB options to keep the socket alive. This issue may have emerged after the MongoDB instance was migrated from MongoHQ to Rackspace to Azure.

**Comment by jaraco** [comment]
I tried enabling socketKeepAlive for our pymongo client

```
diff --git a/pmxbot/storage.py b/pmxbot/storage.py
index 26dc048..bce7dc4 100644
--- a/pmxbot/storage.py
+++ b/pmxbot/storage.py
@@ -102,9 +102,10 @@ class MongoDBStorage(Storage):
        @classmethod
        def _get_collection(cls, uri):
                importlib.import_module('pymongo')
+               client = pymongo.MongoClient(uri, socketKeepAlive=True)
                uri_p = pymongo.uri_parser.parse_uri(uri)
                db_name = uri_p['database'] or 'pmxbot'
-               return pymongo.MongoClient(uri)[db_name][cls.collection_name]
+               return client[db_name][cls.collection_name]
 
 
 def migrate_all(source, dest):
```

But the issue has occurred again, this time apparently with a simple, logged pubmsg.

**Comment by jaraco** [comment]
In c7d8580, I added the ability to pass arbitrary arguments to the MongoClient constructor. I then added the following to the config:

```
database params:
  socketKeepAlive: true
  maxIdleTimeMS: 30000
  socketTimeoutMS: 30000
```

And restarted the bot. It's been connected and responding reliably for over a half-hour now, so I'm hopeful that the maxIdleTimeMS or socketTimeoutMS setting has been sufficient to overcome whatever factor was causing the MongoDB connections to drop.
